cli/src/doctest.rs

```diff
@@ -0,0 +1,417 @@
+}
+
+#[derive(Debug)]
+enum Expected {
+    Value(String),
+    Error(String),
+    None,
+}
+
+impl Expected {
+    /// Parse out an expected outcome from a doctest.
+    ///
+    /// After ignoring whitespace-only trailing lines, we look for the last comment block in the doctest.
+    /// If that comment block has a line starting (modulo whitespace) with "=>", everything following
+    /// the "=>" is the expected value of the doctest.
+    ///
+    /// There are two special cases for tests that are expected to fail:
+    /// - if the "=>" line looks like "=> error: some text", the test is expected to exit with an error,
+    ///   and the error message is supposed to contain "some text".
+    /// - if the "=>" line looks like "=> error", the test is expected to exit with an error (but we
+    ///   don't care what the message is.
+    fn extract(doctest: &str) -> Self {
+        let mut lines: Vec<&str> = doctest.lines().collect();
+
+        // Throw away trailing empty lines.
+        let last_non_empty = lines
+            .iter()
+            .rposition(|line| !line.trim().is_empty())
+            .unwrap_or(0);
+        lines.truncate(last_non_empty + 1);
+
+        let mut expected = Vec::new();
+        for line in lines.iter().rev() {
+            // If we encounter an uncommented line before we find a "=>", there's no expected value
+            // for this test.
+            let Some(commented) = line.trim_start().strip_prefix('#') else {
+                break;
+            };
+
+            if let Some(arrowed) = commented.trim_start().strip_prefix("=>") {
+                // We've found an expected value for the test.
+                if let Some(msg) = arrowed.trim_start().strip_prefix("error:") {
+                    expected.push(msg.trim());
+                    expected.reverse();
+                    return Expected::Error(expected.join("\n"));
+                } else if arrowed.trim() == "error" {
+                    return Expected::Error(String::new());
+                } else {
+                    expected.push(arrowed);
+                    expected.reverse();
+                    return Expected::Value(expected.join("\n"));
+                }
+            } else {
+                expected.push(commented);
+            }
+        }
+
+        Expected::None
+    }
+
+    fn error(&self) -> Option<String> {
+        match self {
+            Expected::Error(s) => Some(s.clone()),
+            _ => None,
+        }
+    }
+}
+
+#[derive(Debug)]
+struct DocTest {
+    input: String,
+    expected: Expected,
+}
+
+struct TestEntry {
+    expected_error: Option<String>,
+    field_name: LocIdent,
+    test_idx: usize,
+}
+
+#[derive(Default)]
+struct TestRegistry {
+    tests: HashMap<Ident, TestEntry>,
+}
+
+impl DocTest {
+    fn new(input: String) -> Self {
+        let expected = Expected::extract(&input);
+        DocTest { input, expected }
+    }
+
+    fn code(&self) -> String {
```

Comment:
I've decided to do this a bit differently, building up the contract application programmatically instead of by modifying the text (which gives slightly weird error messages).